level00/PropertyTests.hs

```diff
@@ -112,7 +113,45 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
 genMyBTreeVal :: MonadGen m => m (Int, Char)
 genMyBTreeVal = liftA2 (,) (Gen.int (Range.linear (-100) 100)) (Gen.enum 'a' 'z')
 
--- We're not ready to write a property test for inserting values into our binary
+-- We can also use property based testing to ensure that when we
+-- implement a typeclass instance, that implementation will comply
+-- with the laws of that typeclass. This is important as often the
+-- laws of a typeclass cannot be expressed in the types.
+--
+-- NOTE: There are _two_ valid 'Eq' instances we could write:
+-- - Structural equality
+-- - Matching set of (key, value) pairs.
+--
+-- Implement the 'Eq' instance for 'MyBTree' that compares for two
+-- trees having a matching set of (key, value) pairs.
+--
+instance (Eq k, Eq a) => Eq (MyBTree k a) where
+  (==) :: MyBTree k a -> MyBTree k a -> Bool
+  (==) a b = toListWithKey a == toListWithKey b
+
+prop_MyBTree_LawfulEqInstance :: Property
+prop_MyBTree_LawfulEqInstance = property $ do
+  (x,y,z) <- forAll $ (,,)
+             <$> genTree genMyBTreeVal
+             <*> genTree genMyBTreeVal
+             <*> genTree genMyBTreeVal
+
+  annotate "Reflexivity: x == x = True"
+  (x == x) === True
+
+  annotate "Symmetry: x == y = y == x"
+  (x == y) === (y == x)
+
+  annotate "Transitivity: if x == y && y == x = True then x == z = True"
+  when (x == y && y == z) $ (x == z) === True
+
+  annotate "Substitutivity: if x == y = True and (f :: (Eq a, Eq b) => a -> b) then f x == f y = True"
+  when (x == y) $ (null x == null y) === True
```

Comment:
This is not a test for the whole property. Add a commentary?

Reply:
It tests for only one function. Dunno what a good general property would look like; can you generate functions from trees to `Int`?

Reply:
Bring in `hedgehog-fn` and have at it ?

Reply:
dooo eeet

Might be worth having some language in there to say "if you're getting along well enough, consider jumping off to level 1 instead of doing literally every exercise"?

Reply:
Can't use hedgehog-fn after initial attempts, it also might push the weirdo-needle too far.

That bonus wording is definitely a good idea though.

Reply:
Made `hedgehog-fn` work, it's a bit crunchy though. Would appreciate more eyes to make it palatable.